### ISSUE TITLE
ClamAV Improvement - Support additional AutoIt script operators

### ISSUE THREAD

**Issue by recvfrom** [recvfrom/contributor#5]
**Overview**
ClamAV can be used to extract AutoIt scripts from compiled EXEs, but in testing against 10,000 samples from VirusTotal, there are some cases where it fails.  Some of this appears to be caused by certain script operators not being recognized by ClamAV.

**Proposal**
Investigate what AutoIt script operators are not supported and then update the ClamAV code to support them. 

**Skillsets Involved**
- Linux SW development in C
- SW Reverse Engineering

**Expected Difficulty**
- Beginner/Easy – Shouldn’t require an existing understanding of AutoIt or too much knowledge of the ClamAV codebase, and the fixes will likely be fairly easy to implement

**Technical Info**
- An example of ClamAV being used for AutoIt script extraction: https://twitter.com/SmugYeti/status/1089919472444624896
- An example of doing AutoIt script extraction via other means: https://0xffff0800.blogspot.com/2019/01/defeating-au3-obfusticated-malware.html
- Getting started with ClamAV development: https://github.com/Cisco-Talos/clamav-faq/blob/master/manual/UserManual/development.md

I found and fixed one of these locally, but haven't committed the fix.  The error ClamAV gave when scanning 77e123e86f43095f4e33a5cadce0fed4bdd9060f78088b3195ca41455e802f71 is as follows:
```
clamscan --debug --verbose --leave-temps 77e123e86f43095f4e33a5cadce0fed4bdd9060f78088b3195ca41455e802f71
...
LibClamAV debug: autoit: found unknown op (57)
LibClamAV debug: autoit: decompilation aborted - partial script may exist
LibClamAV debug: autoit: script extracted to /tmp/clamav-238b9ca9e2f2440cba18c150e382a87f.tmp/autoit.001
```

After comparing the output from autoit.001 to the output from the Exe2Aut decompiler output for the same file, it looked like ClamAV didn't support extracting the ternary operators ('?' and ':').  The fix is:

```
diff --git a/libclamav/autoit.c b/libclamav/autoit.c
index 8d9da58e0..82288fe0b 100644
--- a/libclamav/autoit.c
+++ b/libclamav/autoit.c
@@ -503,7 +503,7 @@ static int ea06(cli_ctx *ctx, const uint8_t *base, char *tmpd) {
   unsigned int files=0;
   char tempfile[1024];
   const char prefixes[] = { '\0', '\0', '@', '$', '\0', '.', '"', '#' };
-  const char *opers[] = { ",", "=", ">", "<", "<>", ">=", "<=", "(", ")", "+", "-", "/", "*", "&", "[", "]", "==", "^", "+=", "-=", "/=", "*=", "&=" };
+  const char *opers[] = { ",", "=", ">", "<", "<>", ">=", "<=", "(", ")", "+", "-", "/", "*", "&", "[", "]", "==", "^", "+=", "-=", "/=", "*=", "&=", "?", ":" };
   struct UNP UNP;
   fmap_t *map = *ctx->fmap;
 
@@ -856,6 +856,8 @@ static int ea06(cli_ctx *ctx, const uint8_t *base, char *tmpd) {
        case 0x54: /* /= */
        case 0x55: /* *= */
        case 0x56: /* &= */
+       case 0x57: /* ? */
+       case 0x58: /* : */
          if (UNP.cur_output+4 >= UNP.csize) {
            uint8_t *newout;
            UNP.csize += 512;
```

From analyzing with 10,000 scripts from VirusTotal there are more of these that we need to investigate and support:
```
found unknown op (0)
found unknown op (84)
found unknown op (a5)
found unknown op (63)
found unknown op (1)
```

**Comment by recvfrom** [comment]
Corresponding files from VT:

found unknown op (57)
 - Hash: 01ae48afd6f2416bd2cf6d3b2220981990f0927b0975ccef396207abd2dbf477
 - Download: https://malshare.com/sample.php?action=detail&hash=70d0e5765b47f4927f544fe8bc7acb80

found unknown op (0)
 - Hash: 0ddaf66568c139e6f8d2fdfdfeb1cfd127ff594a63fb54c4e0e2e689fcd4e9e4
 - Download: https://malshare.com/sample.php?action=detail&hash=ee9d6c8f03f89916cad19e4c2be4058b

found unknown op (84)
 - Hash: 0c66a3db2b517e5f9a300826bcabd06972d1c189e696e74baf64b3dddbd7dbee
 - Download: https://malshare.com/sample.php?action=detail&hash=96e9cf6cfa2ec08262da4f2dde4dafb3

found unknown op (a5)
 - Hash: b34e51f45e49d7826b3d5487afbf356633aa9bc998c6aaf8b77a3a1fda3fe5e8
 - Download: https://malshare.com/sample.php?action=detail&hash=8bdb92ce3005dcf294c3568384272643

found unknown op (63)
 - Hash: a3fbd78b3b2fa782ee913f5bd07443304f7fa64dbff85f94bcb4f990c02f3905
 - Download: https://malshare.com/sample.php?action=detail&hash=753c31552bb9c4251265d4805825fc7a

found unknown op (1)
 - Hash: d18540ba1a8f612ae7faa8d7a3e6c92e1d05f8a4a38187b364016368a32e8f76
 - Download: https://malshare.com/sample.php?action=detail&hash=63e7a3e75df7f8da20ee5e06c7df4559

**Comment by recvfrom** [comment]
When doing development, fork https://github.com/Cisco-Talos/clamav-devel and work off of the dev/0.102 branch

**Comment by recvfrom** [comment]
The fix for `found unknown op (57)` will be included in ClamAV 0.102.  A fix for `unknown op (0)` and `unknown op (1)` will be included in ClamAV 0.103.  The rest of the errors appear to be related to corrupted AutoIt scripts, so no resolution is possible.
